hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestSevenToEightUpgrade.scala

```diff
@@ -102,4 +110,260 @@ class TestSevenToEightUpgrade extends RecordLevelIndexTestBase {
+    client.clean(cleanInstant.get())
+    client.close()
+
+    // ------------------------------------------------------------------
+    // Step 6: Downgrade to version 6.
+    // ------------------------------------------------------------------
+    val downgradedOptions = hudiOptions ++ Map(
+      HoodieWriteConfig.WRITE_TABLE_VERSION.key -> HoodieTableVersion.SIX.versionCode().toString
+    )
+    metaClient = HoodieTableMetaClient.reload(metaClient)
+    new UpgradeDowngrade(metaClient, getWriteConfig(downgradedOptions, basePath), context, SparkUpgradeDowngradeHelper.getInstance)
+      .run(HoodieTableVersion.SIX, null)
+    // assert that metadata table version is 6 and it has been compacted
+    var metadataMetaClient = HoodieTableMetaClient.builder().setConf(metaClient.getStorageConf).setBasePath(basePath + "/.hoodie/metadata").build()
+    assertEquals(HoodieTableVersion.SIX, metadataMetaClient.getTableConfig.getTableVersion)
+    assertEquals(1, metadataMetaClient.getActiveTimeline.getCommitTimeline.getInstants.size())
+    val compactionInstant = metadataMetaClient.getActiveTimeline.getCommitTimeline.filterCompletedInstants().lastInstant().get()
+
+    // ------------------------------------------------------------------
+    // Step 7: Read the table using a predicate on 'price' to force the column stats read path.
+    // This should trigger the error during deserialization of the delete block.
+    // ------------------------------------------------------------------
+    val df = spark.read.format("hudi").load(basePath).filter("price > 15")
+    // assert the result
+    assertEquals(1, df.count())
+    // assert that only id=2 is present
+    assertEquals(1, df.filter("id = 2").count())
+
+    // ------------------------------------------------------------------
+    // Step 8: Do a commit with table version 6 but this time trigger MDT compaction.
+    // ------------------------------------------------------------------
+    val writeConfigsPostDowngrade = hudiOptions ++ Map(
+      HoodieMetadataConfig.COMPACT_NUM_DELTA_COMMITS.key -> "1",
+      HoodieWriteConfig.WRITE_TABLE_VERSION.key -> HoodieTableVersion.SIX.versionCode().toString
+    )
+    val updateDF2 = Seq(
+      (1, "Alice", 3000, 20),  // update Alice
+      (2, "Bob",   3000, 30)   // update Bob
+    ).toDF("id", "name", "ts", "price")
+    updateDF2.write.format("hudi")
+      .options(writeConfigsPostDowngrade)
+      .mode("append")
+      .save(basePath)
+    // assert that metadata table has been compacted
+    metadataMetaClient = HoodieTableMetaClient.reload(metadataMetaClient)
+    val lastCompactionInstant = metadataMetaClient.getActiveTimeline.getCommitTimeline.filterCompletedInstants().lastInstant().get()
+    assertTrue(compareTimestamps(lastCompactionInstant.requestedTime(), GREATER_THAN_OR_EQUALS, compactionInstant.requestedTime()))
+  }
+
+  @Test
+  def testUpgradeDowngradeWithRecordIndex() : Unit = {
```

Comment:
do we need two diff tests. why can't we enable RLI in the other tests only and validate in the same test. 

Trying to avoid more tests if we can piggy back on the same func test to test more scenarios.

Reply:
Is the concern about test time? The RLI issue happens when upgrading. So we start with tv=6 and then upgrade. 
While the metadata compaction scheduling issue happens on commit after downgrade so we start with usual tv=8 and then downgrade.
We can have one test with upgrade followed by downgrade, but the overall time almost remains the same. I thought of keeping the two tests separate as they test two different issues.

Reply:
Btw, I also ensured to do clean and have delete block in metadata table. And ensure that too is compacted correctly.

Reply:
don't we have any other tests that aleady exists for upgrade, downgrade scenario is my question (even if its not the new test you have written). we should have existing tests right?

Reply:
I could not find a test that does upgrade and downgrade with cleaning and compaction. I have merged the two tests into one as discussed above.

Reply:
ok